LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownlinkBasicsStationMessage.cs

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoRaTools.LoRaPhysical
+{
+    using System;
+    using LoRaWan;
+
+    /// <summary>
+    /// JSON of a Downlink message for the Basic Station.
+    /// </summary>
+    public class DownlinkBasicsStationMessage
+    {
+        public string DevEui { get; }
+
+        /// This rx delay control the time we need to wait before sending.
+        /// It is typically 0 for class C devices.
+        public ushort LnsRxDelay { get; }
+
+        public ulong Xtime { get; }
+
+        public uint? AntennaPreference { get; }
+
+        public DataRateIndex DataRateRx1 { get; }
+
+        public DataRateIndex DataRateRx2 { get; }
+
+        public Hertz FrequencyRx1 { get; }
+
+        public Hertz FrequencyRx2 { get; }
+
+#pragma warning disable CA1819 // Properties should not return arrays
+        public byte[] Data { get; }
```

Comment:
Like the warning says, can we not avoid exposing bare arrays?

Reply:
What would you suggest? Shall I put it here in Hex already (we do the conversion in the message atm)? Could you explain the problem that we could have here?

Reply:
I would change this to `ReadOnlyMemory<byte>`, and align the type of `payload` in the constructor accordingly. This way, the whole type is read-only and the byte array cannot be manipulated by anyone else. This change should have minimal impact:

```diff
diff --git a/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownlinkMessage.cs b/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownlinkMessage.cs
index 98f46c8a..5428030f 100644
--- a/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownlinkMessage.cs
+++ b/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownlinkMessage.cs
@@ -29,10 +29,7 @@ namespace LoRaTools.LoRaPhysical
 
         public Hertz FrequencyRx2 { get; }
 
-#pragma warning disable CA1819 // Properties should not return arrays
-        public byte[] Data { get; }
-#pragma warning restore CA1819 // Properties should not return arrays
-
+        public ReadOnlyMemory<byte> Data { get; }
 
         public StationEui StationEui { get; }
 
@@ -45,11 +42,9 @@ namespace LoRaTools.LoRaPhysical
         /// This method is used in case of a response to a upstream message.
         /// </summary>
         /// <returns>DownlinkPktFwdMessage object ready to be sent.</returns>
-        public DownlinkMessage(byte[] payload, ulong xtime, DataRateIndex datrRx1, DataRateIndex datrRx2, Hertz freqRx1, Hertz freqRx2, string devEui, ushort lnsRxDelay = 0, StationEui stationEui = default, uint antennaPreference = 0)
+        public DownlinkMessage(ReadOnlyMemory<byte> payload, ulong xtime, DataRateIndex datrRx1, DataRateIndex datrRx2, Hertz freqRx1, Hertz freqRx2, string devEui, ushort lnsRxDelay = 0, StationEui stationEui = default, uint antennaPreference = 0)
         {
-            if (payload is null) throw new ArgumentNullException(nameof(payload));
             Data = payload;
-
             DevEui = devEui;
             LnsRxDelay = lnsRxDelay;
             AntennaPreference = antennaPreference;
diff --git a/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownstreamSender.cs b/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownstreamSender.cs
index 04ede78c..a014b6eb 100644
--- a/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownstreamSender.cs
+++ b/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownstreamSender.cs
@@ -67,7 +67,7 @@ namespace LoRaWan.NetworkServer.BasicsStation
             // Getting and writing payload bytes
             var pduBytes = message.Data;
             var pduChars = new char[pduBytes.Length * 2];
-            Hexadecimal.Write(pduBytes, pduChars);
+            Hexadecimal.Write(pduBytes.Span, pduChars);
             writer.WriteString("pdu", pduChars);
 
 #pragma warning disable CA5394 // Do not use insecure randomness. This is fine as not used for any crypto operations.
```

Reply:
As discussed it creates more changes required, therefore we will do that in subsequent PR, tracked in #1138